### PR TITLE
Add Nextcloud Notes to exporter plugin

### PR DESCRIPTION
**Pull request by kotnik**
As [requested](https://github.com/koreader/koreader/issues/11522), here is Nextcloud Notes support for exporting notes and highlights.

Fixes https://github.com/koreader/koreader/issues/11522

TODO:

* [ ] Create [wiki](https://github.com/koreader/koreader/wiki/Nextcloud-notes) page mentioned in help

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12301)

### CODE REVIEW

File: plugins/exporter.koplugin/main.lua [hunk 1]
```diff
@@ -102,6 +102,7 @@ local Exporter = WidgetContainer:extend{
         flomo = require("target/flomo"),
         html = require("target/html"),
         joplin = require("target/joplin"),
+        nextcloud_notes = require("target/nextcloud_notes"),
```
Comment (Frenzie) [inline plugins/exporter.koplugin/main.lua:105]: I think it's alphabetical. ;-)

**Review by Frenzie — APPROVED** [review]
Pinging @pazos, nothing jumps out to me besides the nit.

**Comment by pazos** [comment]
@kotnik: thank you very much for your time :)

I would like to have a few days to test the plugin myself. I'm short of time right now.

What I would like is feature parity w/ the joplin exporter but I'm not sure if it's already implemented or even doable (Didn't look at your code or at nextcloud api docs yet). I mean, among others:

- ability to update a note if it was changed.
- ability to ignore a note for a book if the same note exists on the server

This way implies the exporter works as a sync server and any past highlight or annotation is deleted from the server when the clients exports/syncs it after deletion.

Please let me know if that's ok with you. If you prefeer some kind of archive without updating/deleting I think we can make it configurable :)

File: plugins/exporter.koplugin/main.lua (hunk 1 above)
Comment (kotnik) [reply plugins/exporter.koplugin/main.lua:105]: Makes sense, [fixed](https://github.com/koreader/koreader/pull/12301/commits/c873dd5624e9c4470ceee8699b2911004ccad46a). Thanks!

**Comment by kotnik** [comment]
> @kotnik: thank you very much for your time :)
> 
> I would like to have a few days to test the plugin myself. I'm short of time right now.

Thank you for the review. This is, obviously, not urgent at all and I am here to make it through. It's quite fun to code Lua!

> What I would like is feature parity w/ the joplin exporter but I'm not sure if it's already implemented or even doable (Didn't look at your code or at nextcloud api docs yet). I mean, among others:
> 
>     * ability to update a note if it was changed.
> 
>     * ability to ignore a note for a book if the same note exists on the server
> 
> 
> This way implies the exporter works as a sync server and any past highlight or annotation is deleted from the server when the client exports/syncs it after deletion.
> 
> Please let me know if that's ok with you. If you prefeer some kind of archive without updating/deleting I think we can make it configurable :)

Yeah, I looked into Joplin, and this (the first) iteration works by uploading every note/annotation to Nextcloud. If the note from the same title already exists it will be updated (version changes can be configured on the Nextcloud side). Notes are tagged with KOReader tag, which in Nextcloud means they will be put in separate directory, so it is easy to work with them.

I can add an option to update or skip if existing note already exists in Nextcloud. I'd appreciate testing as well, it works fine with my local and remote Nextcloud instances, but I am not super sure how it would work with big number of notes.

**Comment by pazos** [comment]
I'm very sorry, @kotnik. I forgot completely about this PR 😭 

> Yeah, I looked into Joplin, and this (the first) iteration works by uploading every note/annotation to Nextcloud. If the note from the same title already exists it will be updated (version changes can be configured on the Nextcloud side). Notes are tagged with KOReader tag, which in Nextcloud means they will be put in separate directory, so it is easy to work with them.

That's completely fine!

> I can add an option to update or skip if existing note already exists in Nextcloud

Not required, current behaviour works fine.

> . I'd appreciate testing as well, it works fine with my local and remote Nextcloud instances, but I am not super sure how it would work with big number of notes.

I've tested briefly and works fine, but I had to generate a bunch of notes to test. The best way for getting tests for free is to merge this and let hardcore hightlighters to use it.


> TODO:

>> Create [wiki](https://github.com/koreader/koreader/wiki/Nextcloud-notes) page mentioned in help

If you want to add documentation before merging this be my guest ;). Otherwise I'll add some notes after merging.


Again, apologies for the late response.

File: plugins/exporter.koplugin/main.lua [hunk 2]
```diff
@@ -40,7 +40,7 @@ local _ = require("gettext")
 
 -- migrate settings from old "evernote.koplugin" or from previous (monolithic) "exporter.koplugin"
 local function migrateSettings()
-    local formats = { "flomo", "html", "joplin", "json", "memos", "my_clippings", "readwise", "text", "xmnote" }
+    local formats = { "flomo", "html", "joplin", "json", "memos", "my_clippings", "nextcloud_notes", "readwise", "text", "xmnote" }
```
Comment (pazos) [inline plugins/exporter.koplugin/main.lua:43]: AFAICT this plugin cannot export to any other Nextcloud facilities other than "Nextcloud Notes". If so, please use the keyword "nextcloud". References to it on the UI are ok as "Nextcloud notes"

**Review by pazos — CHANGES_REQUESTED** [review]
The PR is totally fine as is:

I'm trying to figure out if my requested changes are some kind of OCD or not. Feedback welcome:

Requested changes:

use "nextcloud" instead of "nextcloud_notes" to refer to code and files, since all the integration with nextcloud, for this particular plugin, will have to happen thought the Nextcloud Notes APIs

**Review by pazos — APPROVED** [review]

**Pull request merged**